From: Hendrik Brueckner <brueckner@redhat.com>
Date: Fri, 11 Dec 2009 17:31:32 -0500
Subject: [s390] dasd: support DIAG access for read-only devices
Message-id: <20091211173901.341720782@redhat.com>
Patchwork-id: 21890
O-Subject: [RHEL6 PATCH 3/5] s390 - dasd: support DIAG access for read-only
	devices
Bugzilla: 546309
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

Description
-----------
If a DASD is attached as read-only under z/VM and it is configured
to be used with the DIAG discipline, then setting the device online
will fail.
The message log will show a message 'DIAG initialization failed (rc=4)'.

The DIAG return code 4 indicates that the initialization was successful
and the device is read-only, but the DASD device driver interprets this
as an error.

The solution is to fix the DIAG device initialization to accept a return
value of 4 as indication of success.

Bugzilla
--------
BZ 546309
https://bugzilla.redhat.com/show_bug.cgi?id=546309

Upstream status of the patch
----------------------------
The patch will be upstream as of kernel version 2.6.33
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=22825ab7693fd29769518a0d25ba43c01a50092a

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Please ACK.

With best regards,

	Hendrik

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/s390/block/dasd_diag.c b/drivers/s390/block/dasd_diag.c
index 4e49b4a..8174ec9 100644
--- a/drivers/s390/block/dasd_diag.c
+++ b/drivers/s390/block/dasd_diag.c
@@ -145,6 +145,15 @@ dasd_diag_erp(struct dasd_device *device)
 
 	mdsk_term_io(device);
 	rc = mdsk_init_io(device, device->block->bp_block, 0, NULL);
+	if (rc == 4) {
+		if (!(device->features & DASD_FEATURE_READONLY)) {
+			dev_warn(&device->cdev->dev,
+				 "The access mode of a DIAG device changed"
+				 " to read-only");
+			device->features |= DASD_FEATURE_READONLY;
+		}
+		rc = 0;
+	}
 	if (rc)
 		dev_warn(&device->cdev->dev, "DIAG ERP failed with "
 			    "rc=%d\n", rc);
@@ -433,16 +442,20 @@ dasd_diag_check_device(struct dasd_device *device)
 	for (sb = 512; sb < bsize; sb = sb << 1)
 		block->s2b_shift++;
 	rc = mdsk_init_io(device, block->bp_block, 0, NULL);
-	if (rc) {
+	if (rc && (rc != 4)) {
 		dev_warn(&device->cdev->dev, "DIAG initialization "
 			"failed with rc=%d\n", rc);
 		rc = -EIO;
 	} else {
+		if (rc == 4)
+			device->features |= DASD_FEATURE_READONLY;
 		dev_info(&device->cdev->dev,
-			 "New DASD with %ld byte/block, total size %ld KB\n",
+			 "New DASD with %ld byte/block, total size %ld KB%s\n",
 			 (unsigned long) block->bp_block,
 			 (unsigned long) (block->blocks <<
-					  block->s2b_shift) >> 1);
+					  block->s2b_shift) >> 1,
+			 (rc == 4) ? ", read-only device" : "");
+		rc = 0;
 	}
 out_label:
 	free_page((long) label);
